### PR TITLE
[#763] New practice exercise `affine-cipher`

### PR DESCRIPTION
**Pull request by jiegillet**
One more exercise.

All files (minus `config.json` and `tests.toml`) were generated automatically from `bin/boostrap_practice_exercise.exs`. Of course, the tests needed a bunch of manual input, but usually at the level of search and replace, so it's quite valuable. 

I tried to be a little bit clever about some things (for example `@spec` generation for the `lib` file) but I hope that contributors understand that they should not trust these guesses too much. Is it better to make them obviously wrong?  
Please test that script by running `elixir bin/boostrap_practice_exercise.exs darts` (or whichever exercise is not done yet, I have not tested for those yet) and inspect the generated files. I'm sure there is room for improvement. 

Now to think of it, should I make a different PR for this?

### CODE REVIEW

File: config.json
Comment (jiegillet) [inline config.json:2466]: Sorry, I ran the checks on a slightly older branch that didn't have the concept tracking. 
I'll remove `enum`. 
Maybe I could add `pipe-operator` instead? `encode` is just one clean pipe.

**Comment by angelikatyborska** [comment]
> Now to think of it, should I make a different PR for this?

Yes, a separate PR with just the bootstrap script would be great so that we don't need to block merging the exercise if the script need improvements or the other way around.

I'll review only the exercise on this PR and wait for a new PR with the script 🙂

File: config.json
Comment (angelikatyborska) [inline config.json:2473]: Pattern matching needs some friends to reach its full potential 😁 
```suggestion
          "pattern-matching",
          "guards",
          "multiple-clause-functions",
```

File: config.json
Comment (angelikatyborska) [inline config.json:2478]: Let's force people to know all conditional expressions and choose the appropriate one on their own (usually those can be interchangeable anyway)
```suggestion
          "if",
          "cond",
          "case",
```

File: config.json
Comment (angelikatyborska) [inline config.json:2480]: The key is a map, so we also need to require maps.

For the code that actually changes the letter to a different one, we need either charlists (your approach) or list comprehensions (alternative approach)
```elixir
      message =
        encrypted
        |> String.replace(@ignored, "")
        |> to_charlist()
        |> Enum.map(fn
          digit when ?0 <= digit and digit <= ?9 -> digit
          char -> Integer.mod(mmi * (char - ?a - b), @alphabet_size) + ?a
        end)
        |> to_string
```

```elixir
      message = String.replace(encrypted, @ignored, "")

      message =
        for << char <- message >>, into: <<>> do
          case char do
            char when ?0 <= char and char <= ?9 -> <<char>>
            char -> <<Integer.mod(mmi * (char - ?a - b), @alphabet_size) + ?a>>
          end
        end
```

```suggestion
          "tuples",
          "maps",
          "list-comprehensions",
          "charlists"
```

File: config.json
Comment (angelikatyborska) [inline config.json:2485]: There is a lot going on in this exercise, so I find it really hard to say what "practices" would be best. I think those make sense.

**Review by angelikatyborska — COMMENTED** [review]
Excellent work as always. You're unstoppable with the new exercises 😁.

I left a request to extend prerequisites and some optional feedback for the stub.

File: exercises/practice/affine-cipher/lib/affine_cipher.ex
Comment (jiegillet) [reply exercises/practice/affine-cipher/lib/affine_cipher.ex:7]: I have no idea how that happened :)

**Review by angelikatyborska — APPROVED** [review]
Looks great 👍 I'll wait a bit before merging to see if @neenjaw wants to add a review too.

**Comment by neenjaw** [comment]
Probably don't have time to look at this today, I trust your judgement @angelikatyborska :D

**Comment by angelikatyborska** [comment]
No worries! I understand that sometimes less review work is better, so I'll just merge and @neenjaw can submit a patch PR in case there's something he really wants changed 😀

**Pull request merged**
